src/ray/raylet/scheduling/cluster_task_manager_test.cc

```diff
@@ -1557,6 +1561,287 @@ TEST_F(ClusterTaskManagerTest, PopWorkerExactlyOnce) {
+    num_callbacks++;
+  };
+  task_manager_.QueueAndScheduleTask(task, &reply, callback);
+  task_manager_.QueueAndScheduleTask(task2, &reply, callback);
+  task_manager_.QueueAndScheduleTask(task3, &reply, callback);
+  pool_.TriggerCallbacks();
+
+  // We shouldn't cap anything for zero cpu tasks (and shouldn't crash before
+  // this point).
+  ASSERT_EQ(num_callbacks, 3);
+}
+
+/// Test that we exponentially increase the amount of time it takes to increase
+/// the dispatch cap for a scheduling class.
+TEST_F(ClusterTaskManagerTest, SchedulingClassCapIncrease) {
+  auto get_unblocked_worker = [](std::vector<std::shared_ptr<MockWorker>> &workers)
+      -> std::shared_ptr<MockWorker> {
+    for (auto &worker : workers) {
+      if (worker->GetAllocatedInstances() != nullptr && !worker->IsBlocked()) {
+        return worker;
+      }
+    }
+    return nullptr;
+  };
+
+  int64_t UNIT = RayConfig::instance().scheduling_class_capacity_interval_ms() * 1e6;
+  std::vector<RayTask> tasks;
+  for (int i = 0; i < 10; i++) {
+    RayTask task = CreateTask({{ray::kCPU_ResourceLabel, 8}},
+                              /*num_args=*/0, /*args=*/{});
+    tasks.emplace_back(task);
+  }
+
+  rpc::RequestWorkerLeaseReply reply;
+  int num_callbacks = 0;
+  auto callback = [&num_callbacks](Status, std::function<void()>, std::function<void()>) {
+    num_callbacks++;
+  };
+  for (const auto &task : tasks) {
+    task_manager_.QueueAndScheduleTask(task, &reply, callback);
+  }
+
+  auto runtime_env_hash = tasks[0].GetTaskSpecification().GetRuntimeEnvHash();
+  std::vector<std::shared_ptr<MockWorker>> workers;
+  for (int i = 0; i < 10; i++) {
+    std::shared_ptr<MockWorker> worker =
+        std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234, runtime_env_hash);
+    pool_.PushWorker(std::static_pointer_cast<WorkerInterface>(worker));
+    pool_.TriggerCallbacks();
+    workers.push_back(worker);
+  }
+  task_manager_.ScheduleAndDispatchTasks();
+
+  ASSERT_EQ(num_callbacks, 1);
+
+  current_time_ns_ += UNIT;
+  ASSERT_FALSE(workers.back()->IsBlocked());
+  ASSERT_TRUE(task_manager_.ReleaseCpuResourcesFromUnblockedWorker(
+      get_unblocked_worker(workers)));
+  task_manager_.ScheduleAndDispatchTasks();
+  pool_.TriggerCallbacks();
+  task_manager_.ScheduleAndDispatchTasks();
+  ASSERT_EQ(num_callbacks, 2);
+
+  // Since we're increasing exponentially, increasing by a unit show no longer be enough.
+  current_time_ns_ += UNIT;
+  ASSERT_TRUE(task_manager_.ReleaseCpuResourcesFromUnblockedWorker(
+      get_unblocked_worker(workers)));
+  task_manager_.ScheduleAndDispatchTasks();
+  pool_.TriggerCallbacks();
+  task_manager_.ScheduleAndDispatchTasks();
+  ASSERT_EQ(num_callbacks, 2);
+
+  // Now it should run
+  task_manager_.ScheduleAndDispatchTasks();
+  current_time_ns_ += UNIT;
+  task_manager_.ScheduleAndDispatchTasks();
+  pool_.TriggerCallbacks();
+  task_manager_.ScheduleAndDispatchTasks();
+  ASSERT_EQ(num_callbacks, 3);
+}
```

Comment:
We should also check the behavior in the other cases that were commented on/fixed earlier:
- check that timer gets reset if any task finishes, not just when all tasks finish
- check that timer is still set when the time between task submissions is long 

It would also be nice to check that the timer doesn't affect scheduling for a different class.